### PR TITLE
WIP (do not merge): "proper" fork containing OlliW's betacopter36dev-v005

### PR DESCRIPTION
**Pull request by kd0aij**
This branch was created by merging code from OlliW's betacopter36dev-v005 into a proper fork of ardupilot. (note that his code was not a proper fork of Ardupilot and that Olli was not interested in learning git)
Its purpose is to facilitate discussion of the addition of support for Olli's to Ardupilot.
I have removed the "NeedRework" label and added "WIP"; also added "do not merge" to the title in hope of avoiding further drama.

### CODE REVIEW

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.h [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+#define AP_BATTMONITOR_UAVCAN_TIMEOUT_MICROS         5000000 // sensor becomes unhealthy if no successful readings for 5 seconds
+
+class AP_BattMonitor_UAVCAN : public AP_BattMonitor_Backend
+{
+public:
+
+    enum BattMonitor_UAVCAN_Type {
+        UAVCAN_CIRCUIT_STATUS = 0,
+        UAVCAN_BATTERY_INFO,
+//OW
```
Comment (amilcarlucas) [inline libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.h]: Remove all //OW lines

File: libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.h [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+        UAVCAN_GENERIC_BATTERY_INFO
+//OWEND
```
Comment (amilcarlucas) [inline libraries/AP_BattMonitor/AP_BattMonitor_UAVCAN.h]: remove all //OWEND lines

File: libraries/AP_Camera/AP_Camera.cpp [hunk 3]
```diff
@@ -11,6 +11,10 @@
 #include <fcntl.h>
 #include <unistd.h>
 #endif
+//OW
+#include "../ArduCopter/Copter.h"
+//extern Copter copter; //not needed then Copter.h is included
+//OWEND
```
Comment (amilcarlucas) [inline libraries/AP_Camera/AP_Camera.cpp]: This one is a NO, NO
library code must not include vehicle code

File: libraries/AP_Camera/AP_Camera.cpp [hunk 4]
```diff
@@ -143,6 +147,10 @@ void AP_Camera::trigger_pic()
         break;
     }
 
+//OW
+    copter.letmeset_trigger_pic(true);
```
Comment (amilcarlucas) [inline libraries/AP_Camera/AP_Camera.cpp]: You need to find a better way to call this

File: ArduCopter/version.h [hunk 5]
```diff
@@ -1,10 +1,47 @@
 #pragma once
 
-#ifndef FORCE_VERSION_H_INCLUDE
```
Comment (amilcarlucas) [inline ArduCopter/version.h]: I think this can not go in, It was probably a badly solved merge conflict ?

File: ArduCopter/version.h [hunk 6]
```diff
@@ -1,10 +1,47 @@
...
+#define FIRMWARE_VERSION 3,6,0,FIRMWARE_VERSION_TYPE_DEV
+
+#ifndef GIT_VERSION
+#define FIRMWARE_STRING THISFIRMWARE
+#else
+#define FIRMWARE_STRING THISFIRMWARE " (" GIT_VERSION ")"
 #endif
 
-#include "ap_version.h"
+/*
```
Comment (amilcarlucas) [inline ArduCopter/version.h]: These bellong in the git comit messages, not in the file

**Comment by kd0aij** [comment]
@amilcarlucas The (initial) intent here is to just get these changes into a proper fork of ardupilot and compiling cleanly. After that is accomplished we can address the stylistic issues and remove any temporary comment lines.

**Comment by lvale** [comment]
thanks @kd0aij 
Next we have to do a git workshop with our friend @olliw

**Comment by OlliW** [comment]
Hey there, it seems you've mentioned the wrong user in your comment. I am not the creator of betacopter36dev-v005.

**Comment by kd0aij** [comment]
@lvale I guess it should be @olliw42...
I forked the  libuavcan and dsdl modules in my own repo and am  down to a few undefined references, but I'm not sure how to get CI to build using my forks of submodules. I could use a git workshop on that topic myself :)

**Comment by lvale** [comment]
@kd0aij Mark we all need a git workshop... :)

**Comment by OlliW** [comment]
Would you be so kind and drop me from this? I am not involved in your project, but keep getting mails about it.

**Comment by lvale** [comment]
@OlliW Would you be so kind to unsubscribe. I can't change it, but you can

screen shot 2017-11-02 at 22 38 44


Sorry for the inconvenience

**Comment by OlliW** [comment]
Oh, thank you for pointing that out; I have no experience using GitHub! Good luck with your project :)

**Comment by kd0aij** [comment]
this is compiling for copter now, but needs the changes @amilcarlucas  pointed out to work for other builds; it will fail for other vehicle types because of course no copter object will be defined
Thanks @amilcarlucas for pointing that out!

@tridge I think the local DSDL file parsing is working correctly now, but haven't looked into the message ID issues yet.

**Comment by olliw42** [comment]
@kd0aij 
THX so much for this. I must admit that I'm a bit caught by surprise and need to readjust as my head (and workbench) is currently full of other stuff. Also, I'm a bit overwhelmed with digesting what's in this PR because I find what I get displayed by git confusing. Pl give me some time. So, just my first 2 cents.

You obviously have found the info in version.h, and got the hang on the //OW //OWEND thing, it looks to me that you caught all modifications.

The changes in RCInput.h/cpp were an attempt to try what was suggested in https://github.com/ArduPilot/ardupilot/issues/6096, but this did not work (not unexpectedly). So, these two files you should IMHO take out of this PR (I guess I should have removed these parts myself). The issue is resolved for the mounts I've added in a more straightforward way.

I'm not sure about the RCOutput_Tap.h/cpp, can't recall that. 

Some of the changes in the UAVCAN part come from a PR of EShamaev, https://github.com/ArduPilot/ardupilot/pull/6527, I've added  so to say half of it to get the circuitstatus message working. Maybe it would be useful to see this PR and this one here as to build on each other.

Remains to touch the "critical" part: dsdl. In order to get this working I've added new UAVCAN data type files to two places within the uavcan namespace. I unfortunately could not do better. However, betacopter36dev-005 is some months old now and since then no better solution turned up, so I have decided few weeks ago to move all my dsdls into a common uavcan subfolder named olliw, to not have them spread around in various places. Such changes affect the data type signature, and this thus breaks compatibility. So far I was planing to release a new STorM32 firmware in about 1-2 weeks, in which this change would become effective (I obviously would then also have provided an updated betacopter). I mention this, because a danger of having to move around the dsdls repeatedly with the accompanying confusions is looming here, which might not be ideal. A somewhat similar situation is with the GenericBatteryInfo message, which I also have planned to move to the olliw subfolder. Any change means that users have to reflash their stuff. I don't know what to suggest here, but  just see a problem here.

That's so far. A last question maybe. If I would want to compile master including this PR myself, what would I have to do? I'm using Github Desktop (windows), and don't find a branch in the ardupilot repository which I would associate with this here.

thx, Olli

(yes, here I'm olliw42 and not OlliW as everywhere else, since OlliW was claimed already when I registered ;))

**Comment by auturgy** [comment]
@olliw42 the git tool of choice for Windows seems to be https://www.sourcetreeapp.com/

**Comment by kd0aij** [comment]
@ollw42 I'd start by forking the ardupilot repo; go here: https://github.com/ArduPilot/ardupilot
and click on the "fork" button at upper right.  You'll need your own fork in order to create pull requests against the ardupilot repo.
I just realized that there are ardupilot-specific instructions here: http://ardupilot.org/dev/docs/where-to-get-the-code.html and that they were probably written by someone with much more experience than I...

I don't know why cloning a repository would damage or replace an existing one. This is the way I would do it using the command line (should be the same in Windows and Linux). I keep all of my git repo's in a folder named git; whenever I want to clone someone else's code, I create a new subfolder there, and then run "git clone" in the new subfolder. 
```
09:45:39 ~/git$ mkdir olliw
09:45:52 ~/git$ cd olliw
09:45:55 ~/git/olliw$ git clone https://github.com/rbeall/ardupilot.git
Cloning into 'ardupilot'...
remote: Counting objects: 200541, done.
remote: Total 200541 (delta 0), reused 0 (delta 0), pack-reused 200540
Receiving objects: 100% (200541/200541), 90.03 MiB | 2.28 MiB/s, done.
Resolving deltas: 100% (146512/146512), done.
Checking connectivity... done.
remote: Counting objects: 200541, done.
09:48:59 ~/git/olliw$ ls
ardupilot
09:49:04 ~/git/olliw$ cd ardupilot/
09:49:07 (master) ~/git/olliw/ardupilot$ git status
On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean
10:23:07 (master) ~/git/olliw/ardupilot$ git remote -v
origin	https://github.com/rbeall/ardupilot.git (fetch)
origin	https://github.com/rbeall/ardupilot.git (push)
```
So now I have cloned Ryan Beall's fork of ardupilot, (pretending it was yours) into git/olliw/ardupilot
To checkout a branch from the kd0aij fork, I'd do this:
first add a remote repo, then fetch and checkout the branch
This puts you in 'detached HEAD' state because you've created a local copy of the branch which is not "tracking" the kd0aij branch. Now give it your own name using "git checkout -b <name>"
and you have your own new branch in your own fork of ardupilot, from which you can generate PRs

```
10:23:36 (master) ~/git/olliw/ardupilot$ git remote add kd0aij https://github.com/kd0aij/ardupilot.git
10:23:44 (master) ~/git/olliw/ardupilot$ git remote -v
kd0aij	https://github.com/kd0aij/ardupilot.git (fetch)
kd0aij	https://github.com/kd0aij/ardupilot.git (push)
origin	https://github.com/rbeall/ardupilot.git (fetch)
origin	https://github.com/rbeall/ardupilot.git (push)
10:23:47 (master) ~/git/olliw/ardupilot$ git fetch kd0aij
...
10:24:14 (master) ~/git/olliw/ardupilot$ git checkout kd0aij/feature-olliw-uavcan 
Note: checking out 'kd0aij/feature-olliw-uavcan'.

You are in 'detached HEAD' state. You can look around, make experimental
changes and commit them, and you can discard any commits you make in this
state without impacting any branches by performing another checkout.

If you want to create a new branch to retain commits you create, you may
do so (now or later) by using -b with the checkout command again. Example:

  git checkout -b <new-branch-name>

HEAD is now at c4c8d05... fix uavcan submodule ref
10:24:26 ((c4c8d05...)) ~/git/olliw/ardupilot$ git checkout -b my_feature_branch
Switched to a new branch 'my_feature_branch'
10:24:43 (my_feature_branch) ~/git/olliw/ardupilot$ 

```

**Comment by olliw42** [comment]
thx for taking so much time to explain all that
sadly, my intention was not to fork or pr or branch, just to get a clone on my pc to be able to compile
gladly, I finally got it done (I think), and could set up Eclipse, as I usually do it (I follow the wiki Building for Pixhawk/PX4 using Eclipse on Windows, which by the way is really great).
sadly, the compile for px4-v2 terminates with an error "fatal error: uavcan/storm32/NodeSpecific.hpp: No such file or directory". 

so, I would think, the dsdl compilation doesn't yet seem to work for me. Is that maybe because I use make?

once I get it compiled I could do some bench tests. I currently can't do flight tests, since my copter is currently equipped with a non-pixhawk flight controller, and this will be so for another couple of weeks. But I think bench tests should help already a lot.

thx, sir

**Comment by kd0aij** [comment]
@olliw42 Sorry, I just assumed you would find it more convenient to have your own fork for future development tasks. Also, it's the only way I know of to contribute directly to the project.

Glad to hear that Eclipse is working for you; I've been an Eclipse user forever... But I'm currently using QtCreator as my code editor and using the "waf" build process for ardupilot (directly from the command line). I find that Qt uses much less RAM than Eclipse and generally has the same code editing features.

And correct, the DSDL compiler is run by the waf build script:
https://github.com/ArduPilot/ardupilot/blob/master/wscript

I didn't even know that script existed till @tridge pointed me to it yesterday, and I've never been very good with makefiles, so I don't know how to fix the make build. I also don't know how to set up Eclipse or Qt to use waf. But, at least on Linux, the command line stuff works well for me when working with ArduPilot, so I only use the full capabilities of Eclipse when I'm working on Java code.

I just tried a build for px4-v2 myself, and I get the same error. But the px4-v3 build is OK...

**Comment by olliw42** [comment]
@kd0aij many thx for these details
I just tried px-v3, but on my side it fails with identical error message
waf ... just looked up the wiki for waf on windows ... I'm going to pass out here
so, unless eventually a make based solution comes up I'm afraid I'm not going to be of much help

:)

**Comment by kd0aij** [comment]
There's definitely a problem with the dsdl generation: I did a "./waf clean" and now many of the generated files are missing. I'll see if I can get that straightened out for a waf build first. Then maybe we can figure out how to modify the makefiles too.

**Comment by olliw42** [comment]
you probably have seen this: 
3rd paragraph here https://discuss.ardupilot.org/t/uavcan-vendor-specific-messages-with-ardupilot/20091
a change in CMakeLists.txt in the folder \modules\uavcan\libuavcan is needed

**Comment by kd0aij** [comment]
@olliw42 I think the waf build is fixed now. To test it, checkout this branch and run (in the top-level ardupilot folder):
```
./waf clean
./waf configure --board=px4-v2
./waf copter
```

**Comment by olliw42** [comment]
@kd0aij 
GREAT
I'm on Windows, no waf for me

btw, I would think that the dsdls files should not go into a folder dsdl/uavcan/XYZ/..., but rather dsdl/XYZ/...

dsdl/uavcan is in my understanding reserved to the standard data set, and - if my assumption that the parser stops the path at dsdl is correct - the new messages would in fact be placed into the standard set. That is currently the signatures are equal to then one places them into the uavcan submodule's dsdl/uavcan folder. Since, so I assume, you had to move them to get it working I would think that there is still one change needed, like the one I mentioned, to get it working with the new messages not placed in the standard name space.

A lot of guessing here, I know, but that's my guess :)

**Comment by kd0aij** [comment]
@olliw42 When I need to build code on Windows, I start up Ubuntu in a VM :)
And your vendor-specific commands are currently in namespaces uavcan::storm32 and uavcan::equipment::power (just like in betacopter). It looks to me like storm32 is a reasonable place to put vendor-specific commands, but I don't see anything wrong with putting vendor-specific commands in the appropriate uavcan::equipment folder either, as long as the datatype IDs are in the right range.

**Comment by olliw42** [comment]
@kd0aij 
I must admit I'm confused as to what the purpose of the endeavor is, and don't know what to respond
http://uavcan.org/Specification/3._Data_structure_description_language/#vendor-specific-data-types
I would have considered it natural that ArduPilot's dsdls would appear in a namespace ardupilot

you should not consider the approach taken in betacopter as a model example ... I choose it since I didn't know better ... and not because I would consider it appropriate (I in fact consider it the contrary...))

**Comment by kd0aij** [comment]
@olliw42 Thanks, my initial intent was to help get support for your uavcan devices into ardupilot master. Another goal is to make it easier to add support for more/new uavcan devices in the future. But at this stage, all I'm trying to do is get things to compile and run properly with minimal changes.

What hardware would you suggest I buy (or build?) in order to test the gimbal control and battery monitor functions?

**Comment by kd0aij** [comment]
@olliw42 I just read some of the uavcan spec you linked above and I see that using "uavcan" for vendor specific stuff is verboten; what is your opinion (now) on using the generic messages for the storm32?
http://uavcan.org/Specification/7._List_of_standard_data_types/#uavcanequipmentcamera_gimbal

I'll rename the AP_UAVCAN/dsdl/uavcan folder to something else when I understand the code better; it looks like it should be simple to support an arbitrary number of vendor-specific namespaces.

**Comment by olliw42** [comment]
@kd0aij 

> what is your opinion (now) on using the generic messages for the storm32?

the generic messages are IMHO unsuitable; I don't consider them for STorM32
I mean, it would not cost anything to add them to the STorM32 code, but they defeat the whole purpose of the STorM32 ... it's like asking to put a pokemon figure on the top of your car's roof, doesn't cost anything, but why should one do it

>  it looks like it should be simple to support an arbitrary number of vendor-specific namespaces.

that's what I would have thought too (without however understanding the build system LOL)

**Comment by kd0aij** [comment]
@olliw42 What hardware would you suggest I buy (or build?) in order to test storm32 gimbal control and battery monitor functions?

**Comment by olliw42** [comment]
@kd0aij oh sorry, I had written a piece of text but seem to have killed it somehow

That's a real problem. Nothing can be bought.

james patterson and mike kelly each own a uc4h power brick, they might help with testing. I don't know what james is doing with his, he may want to pass his over to you. He also has a pcb so could build another one quickly, or send the pcb to you. If you want I also could send you the .brd file privately, so that you could order and build yourself. (I myself don't have spare pcbs or bricks).

with the gimbal it's even more complicated. If things go well I plan to release the Eagle files of the STorM32 v3.x board, which supports CAN, in a week or so. So you could then order pcbs and electronics and build (I also could send you a file privately now if you want). In any case you would need a gimbal mechanics and build a gimbal. (that's how it is with STorM32, it's not commercial).
The larger problem is the STorM32 firmware. I plan to release a new version along with the brd files. However, there is this issue of where the dsdls are going to be placed. I'm in a conflict. On the one hand I do want to release the firmware and do not want to continue waiting indefinitely for the dsdl issue to be resolved (you know, many months have passed...), on the other hand I would thereby set the namespace for a while to come, and potentially in a way incompatible with e.g. this effort here. In some ways it's not going smooth, and I don't see a best approach.

Sorry for not having better news.

**Comment by auturgy** [comment]
G'day Mark,

I'm not in the best position to help right now - my location is 0.8675° 
S, 89.4364° W

which is great for a holiday, not so good for assisting with this PR.  
My hardware is with Peter Barker (cc'd).  It would be awesome if either 
he or Tridge (who sees Peter most weeks - they're both in Canberra) 
could hook it up via sitl to enable you to test, or test locally.

Otherwise, I'll be back in two weeks.

regards,

James

**Comment by kd0aij** [comment]
@olliw42 @auturgy Thanks for the info and offers to help.  Would a UC4H mag/gps node might be sufficient for testing? It looks like all that requires is an F103 devboard: https://www.aliexpress.com/item/FREE-SHIPPING-New-USBASP-USBISP-AVR-Programmer-USB-ATMEGA8-ATMEGA128-Support-Win7-64K/1811780061.html?spm=a2g0s.9042311.0.0.q0nLEW
and a CAN driver...

**Comment by olliw42** [comment]
@kd0aij The gps/mag node would not test any of the new code parts. It should work with master out of the box. So, it could be a good device to start with playing with UAVCAN, if you don't happen to have any other UAVCAN device, but it is probably not of much help for sake of this PR.

**Comment by kd0aij** [comment]
@olliw42 Am I correct that the datatype IDs for your devices should be changed to the range 20,000-21000?
I suggest using a namespace of ap_uavcan for vendor-specific devices in the AP_UAVCAN library.

**Comment by olliw42** [comment]
@kd0aij that's my understanding too. However I find it hilarious that there are 20000 ids for the standard data set but only 1000 for vendor specific ... so I would care only about the lower limit but not the upper limit, or assume 30000. 
IMHO the GenericBatteryInfo message is somewhat different than the storm32 related messages.  The latter are clearly STorM32 specific, while ArduPilot may want to decide to accommodate the GenericBatteryInfo message as its own. In that case I guess it would be you guys who set the id. For the STorM32 messages I would claim the range 28300-283xx. :)

**Comment by EShamaev** [comment]
First, I would like to ask to separate this PR into several smaller ones.
I propose following:
- PR to add vendor specific messages
- PR to add UART ESC
- PR to add STorM32 gimbal
- base PR for battery messages on the existing PR rather than on master

It might be longer to adapt all changes in long run, but that is anyway better and complies to Ardupilot rules for managing PRs.

Second, regarding vendor specific messages:
- I do not see any point in 1093.GenericBatteryInfo.uavcan as it is a stripped down version of 1092.BatteryInfo.uavcan and it does not provide anything new
- 8300.Status.uavcan, 8301.Control.uavcan are already present in 1040.AngularCommand.uavcan and 1044.Status.uavcan. There might be argument only for use of double instead of float, however for a gimbal I see absolutely no limitation on float even for a high-perfomance gimbals
- 8302.NodeSpecific.uavcan and 8303.NodeSpecificAck.uavcan are a mystery
- 8304.Statistics.uavcan - could not find ay use of that at all

**Comment by olliw42** [comment]
hey

I'm responding with also having yesterday's conversation among you, Pavel, and Francisco on the ArduPilot/CANBUS gitter channel in mind. Since it's related to this PR. And it's unexpectedly absurd.

1. You guys, in the mentioned conversation as well as in several other places, repeatedly make statements like "OlliW wants", where you however obviously have no clue what my intentions are, but take your speculations as correct. This is most inappropriate. You can't put your words in someone elses mouth, and then get in rage about them. So, please read up what I've said or done, or ask. Thx.

2. I consider it inappropriate that a standard has been created which explicitly allows for vendor-specifc messages, but that whenever a vendor wants to make use of that part of the standard gets to hear a litany of false arguments. This is probably mostly directed at Pavel. Note: Francisco pointed out that inconsistency. Please clarify once and for all the situation as regards the vendor-specifc messages - and adapt the standard accordingly.

As the standard currently stands, they are allowed, which implies that it's in the vendor's sole discretion to construct the messages. Period.

3. "as it breaks all possible rules of both Ardupilot codebase and UAVCAN": I think you seriously want to reconsider that statement. Hint: Note that there is a STorM32 Mavlink mount, with specifics to STorM32, there is a STorM32 serial mount, using very specifc STorM32 message (and none was introduced by me but ArduPilot), there is a Alexmos Mount, using very specific Alexmos messages. As regards UAVCAN, see point 2. In view of these facts it appears quite far fetched to make the above claim.

4. As regards 1093.GenericBatteryInfo.uavcan: None of the standard messages can do the job. I find it actually strange that there can be a discussion about it. You can find the use of that message here: https://discuss.ardupilot.org/t/uavcan-for-hobbiests/16464/59 and following post.

5. As regards 8300.Status.uavcan, 8301.Control.uavcan. Pavel himself has acknowledged that the gimbal standard data types are inappropriate. It's a mystery to me why he didn't make that clear in the yesterday's gitter conversation. Note: Francisco did. Hint: it's float16.

6. As regards 8302.NodeSpecific.uavcan, 8303.NodeSpecificAck.uavcan, 8304.Statistics.uavcan: See point 2. In case you think they have no use, you may watch the UAVCAN videos on my youtube channel, olliw42.

Don't get me wrong, I'm very happy to discuss all that. But it should be with conciousness. What I see lately going on is not along this. And it's not good, and no fun.

**Comment by EShamaev** [comment]
1. There is https://github.com/ArduPilot/ardupilot/pull/7208 and I do not mind adding vendor specific messages at all. If comunity decides to accept messages and their handling in AP - let it be. I actually have a few vendor specific messages myself, but never took it out to public. And I will benefit from support of vendor specific messages during compilation.
2. I will not start guessing the "hints" you provide. If there is something specific to say - do it.
3. I asked you about the resolution in float16 quaternions and did not get an answer. However as I stated in other thread - I agree that the resolution could be better even with the current size of message. But I am really against Euler angles and vote for quaternions.
4. As for 8302.NodeSpecific.uavcan, 8303.NodeSpecificAck.uavcan, 8304.Statistics.uavcan I am looking in this PR and see nothing for those messages. Maybe there are other sources like your YouTube channel, but that is definetely out of scope of this PR.
5. For "as it breaks all possible rules of both Ardupilot codebase" - I will not reconsider this. First, you took the phrase out of the context, and as I noticed you do such things regulary. Second, I learned the rules hard way, with some of my first PRs being rejected at all. In this PR there are 4 absolutely different things mixed together: addition of: vendor specific messages, ESC, Gimbal and Battery. All I asked was to separate it in 4 PRs each addressing own issue. And Mark understood my concern and clearly asked to go step by step starting with the https://github.com/ArduPilot/ardupilot/pull/7208

**Comment by olliw42** [comment]
to 1.: good
I wonder what the discussion then was for
to 2.: accepted. I should have just ended the paragraph before that. It wasn't meant as offense but as a pointer, since it was discussed before at several places, and you were talking about double vs float, and not float16. Float is perfectly fine, I'm not happy with float16.
to 3.a: resolution was discussed before at other places. I don't see where you were asking me for it here.
to 3.b: I nowhere said I would be pro Euler and against quaternion. Never opposed it.
to 4.: libraries/AP_UAVCAN/AP_UAVCAN.h, libraries/AP_UAVCAN/AP_UAVCAN.cpp, libraries/AP_Mount/AP_Mount_STorM32_UAVCAN.cpp.
to 5: pl note that this is not my PR.

in few lines you held me accountable three times ("I asked you about the resolution in float16 quaternions and did not get an answer", "As for 8302.NodeSpecific.uavcan ... I ... see nothing for those messages", "In this PR there are 4 absolutely different things mixed together ... all I asked was to separate it in 4 PRs") for things I'm not responsible for. This is not what I call appropriate.

**Comment by EShamaev** [comment]
It's clear that it's not your PR (but I thought that you are behind it, but if you are not I am sorry that I got you into the discussion) and also it's clear that my answers are to your statements in previous message. 

For 8302.NodeSpecific.uavcan, 8303.NodeSpecificAck.uavcan, 8304.Statistics.uavcan I checked again this PR and I still do not see a single use of that in code. So I am asking author of the PR (and not OlliW) why they are needed.
Also the question on separating the PR into 4 is for author of this PR and not for OlliW.

**Comment by kd0aij** [comment]
@EShamaev I've edited the title and labels and added a description of this branch and its purpose in the first comment. This PR was created only to make this branch available for discussion. Sorry if that wasn't clear from the start.

**Comment by kd0aij** [comment]
@olliw42 Your power brick works in this branch with the hacks in the last 2 commits above. Can you see what the problem is?

**Comment by kd0aij** [comment]
Power brick (http://www.olliw.eu/2017/uavcan-for-hobbyists/#chapterpowerbrick) is working properly;
but need to change Olli's GenericBatteryInfo message ID to the vendor-specific range

**Comment by kd0aij** [comment]
mavproxy console display of battery voltage (actual input 8.266V) and current (no load)

mavproxy_power_brick

**Comment by marcdornan** [comment]
Mark -- I have one of Olli's mag/gps and Powerbricks. I am willing to help test this fork but first I need to get a working betacopter as a baseline which may take a week. And Olli does want this stuff to fly and not sit in a drawer. That said, if it is helpful I would be willing to send them to you for a period so you could more expediently work on this (you are in the US, right?). I could put them on a test rig (just the FC and peripherals, not flyable) I have with a Pixracer and ship it to you. Let me know.

Or is there way you could log into a test set up here?

Let me know if I can help.

**Comment by kd0aij** [comment]
@marcdornan Thanks, but I now have a power brick myself. and that should be sufficient for testing. You could help by trying to replicate my success on this branch as reported above: https://github.com/ArduPilot/ardupilot/pull/7167#issuecomment-349079710

**Comment by marcdornan** [comment]
Ok. As soon as I have a baseline of Olli’s binaries working I will then
work on this branch. I really want to ultimately put all of this in a
Plane, if at all possible.

**Comment by auturgy** [comment]
Suggest closing: overtaken by events, and probably better to open a new PR off a more recent branch.

**Comment by kd0aij** [comment]
@auturgy agreed, thanks.

**Pull request closed without merging**